docs(CSS): Allow more usage of multi-keyword `display` values (#30831)
mdn/content: files/en-us/web/css/display/multi-keyword_syntax_of_display/index.md

---
title: Using the multi-keyword syntax with CSS display
slug: Web/CSS/display/multi-keyword_syntax_of_display
page-type: guide
---

{{CSSRef}}

[CSS Display Module Level 3](https://drafts.csswg.org/css-display/) introduced a multi-keyword syntax for the values you can use with a CSS [`display`](/en-US/docs/Web/CSS/display) property. This guide explains the multi-keyword syntax.

> > **Note:** multi-keyword syntax may also be referred to as "two-value syntax" or "multi-value syntax."

Author: done in a68e7c227fd67ea1ceb2f10baa8a50c1959dd153

There are mappings for all of the existing values of `display`; the most common ones are listed in the table below. To see a full list take a look at the table found in the [`display` property specification](https://drafts.csswg.org/css-display/#display-value-summary).

> | Single value   | Current value      |

Author: >i am leaning toward single / multi

I think that's a good call

## display: block flow-root and display: inline flow-root

Regarding how this multi-value syntax helps clarify CSS layout, we can look at some values in the table above that might be less familiar. The multi-keyword `display: block flow-root` maps to a single value; `display: flow-root`. This value's only purpose is to create a new [Block Formatting Context](/en-US/docs/Web/Guide/CSS/Block_formatting_context) (BFC). A BFC ensures that everything inside your box stays inside, and things outside the box cannot intrude into it.

In the example below, four `<div>` elements demonstrate how display values affect formatting contexts.
The first `<div>` element with the demo controls is hidden so we can focus on the elements that follow instead.
The elements that we should focus on are the "parent", "child", and "sibling" `<div>` elements which you can differentiate by their IDs.

What's notable about this layout is that there is no content between the parent and child elements, and the child element has a top margin applied.

> You might expect that the top margin of the child element gives it a positioning somewhere in the middle of the parent element, but what happens instead is something called [margin collapse](/en-US/docs/Web/CSS/CSS_box_model/Mastering_margin_collapsing).

Author: sure, let's do that. Editing your suggestion so I can apply it directly.

In this case, the margin of the child element extends well above the parent's bounding box and pushes the parent element further down the page.
This is easier to see if you inspect the box model of the child element [in your browser's developer tools](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#use_browser_devtools_to_view_the_box_model).

> Choose the options in the `<select>` element to see the effect of different `display` values.

Reviewer: ```suggestion
Change the selected option in the `<select>` element to see the effect of different `display` values.
```

You can use any value with `flow-root` to create a new formatting context, giving the child element a margin relative to its parent and avoiding the margin collapse.

> Changing between `display: flow-root` and `display: block flow-root` will achieve the same effect as the single `flow-root` value.

Reviewer: ```suggestion
Changing between `display: flow-root` and `display: block flow-root` will achieve the same effect as the single-value `flow-root` keyword.
```

Reviewer 2: Não se fala nada inteligível?

  var parentDiv = document.getElementById("parent");
  var siblingDiv = document.getElementById("sibling");
  var displayType = document.getElementById("displayType").value;

  parentDiv.style.display = displayType;
  siblingDiv.style.display = displayType;
}
```

```css hidden
#controls {
  padding: 1rem;
  outline: 2px dashed black;
  background-color: white;
  color: black;
}
body {
  margin: 10px;
  font-family: sans-serif;
  color: white;
}
```

```css
div {
  outline: 2px solid black;
  background-color: cornflowerblue;
  display: block;
  margin-bottom: 40px;
}

#parent {
  background-color: aquamarine;
  height: 140px;
}

#child {
  margin-top: 60px;
  outline: 2px solid red;
}

#sibling {
  background-color: mediumslateblue;
}
```

```html hidden
<div id="controls">
  <label for="displayType">display:</label>
  <select id="displayType" onchange="changeDisplayType()">
    <option value="block">block</option>
    <option value="flow-root">flow-root</option>
    <option value="block flow-root">block flow-root</option>
    <option value="inline">inline</option>
    <option value="inline flow-root">inline flow-root</option>
  </select>
</div>
```

```html
<div id="parent">

>   <div id="child">
>     <p>This para is inside the child element.</p>
>   </div>
> </div>
> <div id="sibling">
>   <p>This is a sibling.</p>
> </div>

Reviewer: ```suggestion
  <div id="child">
    <p>The #child element (nested in the #parent).</p>
  </div>
</div>
<div id="sibling">
  <p>The #sibling (to the #parent)</p>
</div>
```
if you think this helps make it easier to understand

## display: block flow-root and display: inline flow-root

In terms of how this multi-value syntax helps clarify CSS layout, we can take a look at a couple of values in the table that might seem less familiar. The multi-keyword `display: block flow-root` maps to a single value; `display: flow-root`. This value's only purpose is to create a new [Block Formatting Context](/en-US/docs/Web/Guide/CSS/Block_formatting_context) (BFC). A BFC ensures that everything inside your box stays inside it, and things from outside the box cannot intrude into it. When `flow-root` was added as a `display` value, a common use case for creating a new BFC was to contain floats and avoid the need for clearfix hacks.

> In the example below, we have a floated item inside a container. The float is contained by the bordered box, which wraps it and the text alongside. If you remove the line `display: flow-root` then the float will poke out of the bottom of the box. You can replace `display: flow-root` with the multi-value `display: block flow-root`, which will achieve the same as the single `flow-root` value.

Author: Yeah good idea (do we want to avoid floats in general?)

> In the example below, we have a floated item inside a container. The float is contained by the bordered box, which wraps it and the text alongside. If you remove the line `display: flow-root` then the float will poke out of the bottom of the box. You can replace `display: flow-root` with the multi-value `display: block flow-root`, which will achieve the same as the single `flow-root` value.

Author: done in 888bc1a5c0d11002cf62277a3cb33da642e2eca3

```js hidden
function changeDisplayType() {
  var parentDiv = document.getElementById("parent");
  var siblingDiv = document.getElementById("sibling");
  var displayType = document.getElementById("displayType").value;

  parentDiv.style.display = displayType;
  siblingDiv.style.display = displayType;
}
```

```css hidden
#controls {
  padding: 1rem;
  outline: 2px dashed black;
}
body {
  margin: 10px;
  font-family: sans-serif;
}
```

```css
div,
p {
  outline: 2px solid black;
  background-color: cornflowerblue;
  display: block;
  margin-bottom: 2rem;
}

#parent {
  background-color: oldlace;
  min-height: 2rem;
}

#child {
  margin-top: 4rem;
  outline: 2px dashed red;
}

#sibling {
  background-color: lavender;
}
```

```html hidden
<div id="controls">
  <label for="displayType">display:</label>
  <select id="displayType" onchange="changeDisplayType()">
    <option value="block">block</option>
    <option value="flow-root">flow-root</option>
    <option value="block flow-root">block flow-root</option>
    <option value="inline">inline</option>
    <option value="inline flow-root">inline flow-root</option>
  </select>
</div>
```

```html
<div id="parent">

>   <p id="child">The #child paragraph (nested in the #parent).</p>
> </div>
> <p id="sibling">The #sibling paragraph (to the #parent)</p>

Reviewer: ```suggestion
  <p id="child">The #child paragraph (nested in #parent).</p>
</div>
<p id="sibling">The #sibling paragraph (sibling of #parent)</p>
```

## display: block flow-root and display: inline flow-root

Regarding how this multi-value syntax helps clarify CSS layout, we can look at some values in the table above that might be less familiar. The multi-keyword `display: block flow-root` maps to a single value; `display: flow-root`. This value's only purpose is to create a new [Block Formatting Context](/en-US/docs/Web/Guide/CSS/Block_formatting_context) (BFC). A BFC ensures that everything inside your box stays inside, and things outside the box cannot intrude into it.

> In the example below, four `<div>` elements demonstrate how display values affect formatting contexts.

Reviewer: ```suggestion
In the example below, two `<p>` elements, one inside a `<div>` demonstrate how display values affect formatting contexts.
```

Reviewer 2: Quem é o pai?

The first `<div>` element with the demo controls is hidden so we can focus on the elements that follow instead.

> The elements that we should focus on are the "parent", "child", and "sibling" `<div>` elements which you can differentiate by their IDs.

Reviewer: ```suggestion
The elements that we should focus on are the "parent", "child", and "sibling" `<div>` and `<p>` elements which you can differentiate by their IDs.
```

In this case, the margin of the child element extends well above the parent's bounding box and pushes the parent element further down the page.
This is easier to see if you inspect the box model of the child element [in your browser's developer tools](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#use_browser_devtools_to_view_the_box_model).

Choose the options in the `<select>` element to see the effect of different `display` values.
You can use any value with `flow-root` to create a new formatting context, giving the child element a margin relative to its parent and avoiding the margin collapse.
Changing between `display: flow-root` and `display: block flow-root` will achieve the same effect as the single `flow-root` value.

```js hidden
function changeDisplayType() {
  var parentDiv = document.getElementById("parent");
  var siblingDiv = document.getElementById("sibling");
  var displayType = document.getElementById("displayType").value;

  parentDiv.style.display = displayType;
  siblingDiv.style.display = displayType;
}
```

```css hidden
#controls {
  padding: 1rem;
  outline: 2px dashed black;

>   background-color: white;
>   color: black;

Reviewer: ```suggestion
```
no longer needed.

}
body {
  margin: 10px;
  font-family: sans-serif;

>   color: white;

Reviewer: ```suggestion
```
changing colors because the text disappears when the background does.

}
```

```css
div {
  outline: 2px solid black;
  background-color: cornflowerblue;
  display: block;

>   margin-bottom: 40px;

Reviewer: ```suggestion
  margin-bottom: 2rem;
```

}

#parent {

>   background-color: aquamarine;
>   height: 140px;
> }
> 
> #child {
>   margin-top: 60px;
>   outline: 2px solid red;
> }
> 
> #sibling {
>   background-color: mediumslateblue;

Reviewer: ```suggestion
  background-color: oldlace;
  min-height: 2rem;
}

#child {
  margin-top: 4rem;
  outline: 2px dashed red;
}

#sibling {
  background-color: lavender;
```
